src/jdk.httpserver/share/classes/sun/net/httpserver/ServerImpl.java

```diff
@@ -233,12 +234,9 @@ public void stop (int delay) {
         terminating = true;
         try { schan.close(); } catch (IOException e) {}
         selector.wakeup();
-        long latest = System.currentTimeMillis() + delay * 1000;
-        while (System.currentTimeMillis() < latest) {
+        long latest = System.nanoTime() + Duration.ofSeconds(delay).toNanos();
+        while (activeExchanges() > 0 && System.nanoTime() < latest) {
```

Comment:
Also the arithmetic is dubious since `System.nanoTime() + Duration.ofSeconds(delay).toNanos();` could overflow. You never want to compare `val1 < val2` - use `val2 - val1 > 0` instead.